### PR TITLE
Add support for generic list constants

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37105 [JIT] remove dominated guards of functional values
* #37012 optimize is_float_point calls
* **#36953 Add support for generic list constants**
* #36829 [JIT] Add Lower Tuples Call & Run remove mutation after list unrolling
* #36828 add post unroll optimizations
* #36758 Add size peephole optimziation
* #36757 disable peephole optimizations that require alias db

Add support for generic lists as a constant. generic dicts & tuples are already implemented. This is a pretty common pattern and cuts down on the number of non-tensor nodes executed in interpolate tests.

Differential Revision: [D21160761](https://our.internmc.facebook.com/intern/diff/D21160761)

### CODE REVIEW

File: aten/src/ATen/core/ivalue.cpp [hunk 1]
```diff
@@ -287,12 +287,15 @@ std::ostream& printMaybeAnnotatedList(
     std::ostream& out,
     const IValue& the_list,
     IValueFormatter formatter) {
-  if (the_list.toListRef().size() == 0) {
-    out << "annotate(" << the_list.type()->python_str() << ", [])";
+
+  if (the_list.toListRef().size() == 0 || elementTypeCanBeInferredFromMembers(the_list.type())) {
```
Comment (Krovatkin) [inline aten/src/ATen/core/ivalue.cpp:291]: at a very quick glance, shouldn't this be `!elementTypeCanBeInferredFromMembers`, like we should annotate when we can **not** infer the type from members?

**Review by Krovatkin — COMMENTED** [review]
looks fine, i got a bit confused about the annotation part though :-(

File: aten/src/ATen/core/ivalue.cpp (hunk 1 above)
Comment (eellison) [reply aten/src/ATen/core/ivalue.cpp:291]: good catch. that's where test failures are coming from now

**Comment by eellison** [comment]
@BowenBao do you or someone else on ONNX mind taking a look ? i debugged and fix some ONNX callsites but the remaining errors i'm not sure.

**Comment by BowenBao** [comment]
```
Apr 21 17:52:00 test/onnx/test_pytorch_onnx_caffe2.py::TestCaffe2BackendEmbed_opset9::test_mnist Fatal Python error: Segmentation fault
```

I'm having a look.

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -773,11 +772,11 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
                     return g.op("Constant", value_t=n["value"])
                 if n.kindOf("value") == "s":
                     return g.op("Constant", value_s=n["value"])
-                elif n.kindOf("value") == "is":
-                    value = torch.stack([torch.tensor(v) for v in n["value"]]) if n["value"] else []
+                elif n.output().type().isSubtypeOf(ListType.ofInts()):
+                    value = torch.stack([torch.tensor(v) for v in n.output().toIValue()])
```
Comment (BowenBao) [inline torch/onnx/utils.py:776]: empty list not valid input for `torch.stack`
```suggestion
                    vals = n.output().toIValue()
                    value = torch.stack([torch.tensor(v) for v in vals]) if len(vals) > 0 else []
```

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -773,11 +772,11 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
...
                     return g.op("Constant", value_t=value)
-                elif n.kindOf("value") == "fs":
-                    value = torch.stack([torch.tensor(v) for v in n["value"]]) if n["value"] else []
+                elif n.output().type().isSubtypeOf(ListType.ofFloats()):
+                    value = torch.stack([torch.tensor(v) for v in n.output().toIValue()])
```
Comment (BowenBao) [inline torch/onnx/utils.py:779]: maybe merge this branch with `Ints` branch? the above change applies here as well to handle empty list case.

**Comment by BowenBao** [comment]
Commented inline, test locally that should resolve the caffe2 mnist failure.

File: torch/onnx/utils.py (hunk 3 above)
Comment (eellison) [reply torch/onnx/utils.py]: I don't think this code path is executed for the caffe2 mnist failure, bc I uploaded it with a typo and the typo wasn't hit but the seg fault still happened

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -773,11 +772,9 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
...
                 if n.kindOf("value") == "s":
                     return g.op("Constant", value_s=n["value"])
-                elif n.kindOf("value") == "is":
-                    value = torch.stack([torch.tensor(v) for v in n["value"]]) if n["value"] else []
-                    return g.op("Constant", value_t=value)
-                elif n.kindOf("value") == "fs":
-                    value = torch.stack([torch.tensor(v) for v in n["value"]]) if n["value"] else []
+                elif n.output().type().isSubtypeOf(ListType.ofInts()) or output().type().isSubtypeOf(ListType.ofFloats()):
+                    vals = n.output().toIValue()
+                    value = torch.stack([torch.tensor(v)] for v in vals) if len(vals) else []
```
Comment (BowenBao) [inline torch/onnx/utils.py:777]:
```suggestion
                    value = torch.stack([torch.tensor(v) for v in vals]) if len(vals) else []
```

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -92,24 +92,31 @@ def export(model, args, f, export_params=True, verbose=False, training=None,
...
-                           .insertBefore(node).output()
-                          for t in node['value']]
-                lc = (g.create("prim::ListConstruct", inputs)
-                      .insertBefore(node)
-                      .output()
-                      .setType(ListType.ofTensors()))
-                node.output().replaceAllUsesWith(lc)
-
+        if _is_constant_tensor_list(node):
+            inputs = [g.create("prim::Constant").t_('value', t)
```
Comment (BowenBao) [inline torch/onnx/utils.py:113]:
```suggestion
            inputs = [g.create("prim::Constant").t_('value', t).insertBefore(node).output()
                      if t is not None
                      else g.create("prim::Constant").insertBefore(node).output().setType(NoneType.get())
                      for t in node.output().toIValue()]
```

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -92,24 +92,31 @@ def export(model, args, f, export_params=True, verbose=False, training=None,
...
 
 
+def _is_constant_tensor_list(node):
+    if node.kind() != "prim::Constant":
+        return False
+    output_type = node.output().type()
+    if output_type.isSubtypeOf(ListType.ofTensors()):
+        return True
+    if output_type.isSubtypeOf(ListType(OptionalType.ofTensor())):
+        vals = node.output().toIValue()
```
Comment (BowenBao) [inline torch/onnx/utils.py:102]:
```suggestion
        return True
    return False
```

File: torch/onnx/utils.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 from torch._six import string_classes
 from torch.jit import _unique_state_dict
 from torch.onnx import ONNX_ARCHIVE_MODEL_PROTO_NAME, ExportTypes, OperatorExportTypes, TrainingMode
-from torch._C import ListType, _propagate_and_assign_input_shapes, _assign_output_shapes, _check_onnx_proto
+from torch._C import ListType, OptionalType, _propagate_and_assign_input_shapes, _assign_output_shapes, _check_onnx_proto
```
Comment (BowenBao) [inline torch/onnx/utils.py:21]:
```suggestion
from torch._C import ListType, OptionalType, NoneType, _propagate_and_assign_input_shapes, _assign_output_shapes, _check_onnx_proto
```

**Review by BowenBao — COMMENTED** [review]
Several inline suggestions to handle List of optional tensors case where there might be None in the list.

File: torch/onnx/utils.py [hunk 8]
```diff
@@ -92,24 +92,31 @@ def export(model, args, f, export_params=True, verbose=False, training=None,
...
-                           .insertBefore(node).output()
-                          for t in node['value']]
-                lc = (g.create("prim::ListConstruct", inputs)
-                      .insertBefore(node)
-                      .output()
-                      .setType(ListType.ofTensors()))
-                node.output().replaceAllUsesWith(lc)
-
+        if _is_constant_tensor_list(node):
+            inputs = [g.create("prim::Constant").t_('value', t).insertBefore(node).output()
```
Comment (BowenBao) [inline torch/onnx/utils.py:111]:
```suggestion
            inputs = []
            for t in node.output().toIValue():
                n = g.create("prim::Constant").insertBefore(node)
                if t is not None:
                    n.t_('value', t).output().inferTypeFrom(t)
                else:
                    n.output().setType(NoneType.get())
                inputs.append(n.output())
```

**Review by BowenBao — COMMENTED** [review]
Update to have created `prim::Constant` inherit shape and dtype from tensor.

**Pull request closed without merging**